src/persist-txn/src/txn_read.rs

```diff
@@ -315,33 +309,78 @@ pub(crate) struct DataSubscribe<T> {
     pub(crate) remap: DataRemapEntry<T>,
 }
 
+/// An active subscription of [`DataRemapEntry`]s for a data shard.
+#[derive(Debug)]
+pub struct DataSubscription<T: Timestamp + Lattice + Codec64> {
+    /// Metadata and current [`DataRemapEntry`] for the data shard.
+    subscribe: DataSubscribe<T>,
+    /// Channel to send [`DataRemapEntry`]s.
+    tx: mpsc::UnboundedSender<DataRemapEntry<T>>,
+}
+
 impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64> DataSubscribeBlocked<T> {
     /// Unblocks the snapshot portion of this subscribe. See
     /// [`DataSnapshot::unblock_read`].
     ///
-    /// Returns `None` if the upper of the data shard is the empty antichain.
+    /// Returns a [`DataSubscribe`] and a future that must be awaited to
+    /// unblock that snapshot. The returned subscribe is `None` if the upper
+    /// of the data shard is the empty antichain.
+    #[must_use]
     #[instrument(level = "debug", fields(shard = %self.0.data_id, ts = ?self.0.as_of, empty_to = ?self.0.empty_to))]
-    pub(crate) async fn unblock_subscribe<K, V, D>(
+    pub(crate) fn unblock_subscribe<'a, K, V, D>(
         self,
         data_write: WriteHandle<K, V, T, D>,
-    ) -> Option<DataSubscribe<T>>
+    ) -> (Option<DataSubscribe<T>>, BoxFuture<'a, ()>)
     where
-        K: Debug + Codec,
-        V: Debug + Codec,
+        K: Debug + Codec + Send + Sync,
+        V: Debug + Codec + Send + Sync,
         D: Semigroup + Codec64 + Send + Sync,
     {
-        let snapshot = self.0;
-        let empty_to = snapshot.unblock_read(data_write).await;
-        let Some(empty_to) = empty_to.into_option() else {
-            return None;
-        };
-        Some(DataSubscribe {
-            data_id: snapshot.data_id,
+        debug!(
+            "unblock_subscribe latest_write={:?} as_of={:?} for {:.9}",
+            self.0.latest_write,
+            self.0.as_of,
+            self.0.data_id.to_string()
+        );
+
+        if data_write.shared_upper().is_empty() {
```

Comment:
Is this just an optimization or is it structural? If the former, I'm not sure it's worth the complexity

Reply:
I mainly kept this because that's how the old version is structured. However, my understanding is that this should just be an optimization. It's possible that the upper moves to the empty antichain after this method call and we should be correctly handling that. I've pushed a version with this removed and we'll see if things blow up.